liota/dccs/iotcc.py

```diff
@@ -713,10 +783,43 @@ def on_response(msg):
                     self.prop_list = json_msg["body"]["propertyList"]
                 else:
                     log.info("Waiting for getting properties")
-                    on_response(self.recv_msg_queue.get(True, timeout))
+                    on_response(self.get_prop_resp_q.get(True, timeout), get_prop_resp_q)
             except:
                 log.exception("Exception while getting properties")
 
-        self.comms.send(json.dumps(self._get_properties(self._next_id(), resource_uuid)))
-        on_response(self.recv_msg_queue.get(True, timeout))
+        transaction_id = self._next_id()
+        req = Request(transaction_id, get_prop_resp_q)
+        log.debug("Updating get properties response queue for transaction_id:{0}".format(transaction_id))
+        with self.req_ops_lock:
+            self.req_dict.update({transaction_id: req})
+        self.comms.send(json.dumps(self._get_properties(transaction_id, resource_uuid)))
+        on_response(get_prop_resp_q.get(True, timeout), get_prop_resp_q)
         return self.prop_list
+
+    def _dispatch_recvd_msg(self):
+        log.debug("Dispatching received messages from IOTCC")
+
+        while True:
+            try:
+                # block until there is an item available
+                msg = self.recv_msg_queue.get(True)
+                log.debug("Received msg: {0}".format(msg))
+                json_msg = json.loads(msg)
+                log.debug(
+                    "Processing msg: type:{0} transaction_id:{1}".format(json_msg["type"], json_msg["transactionID"]))
+                # search matched request in request dictionary
+                # assume transaction_id will be unique in one process
+                log.debug("Dictionary keys:{0}".format(self.req_dict.keys()))
+                req = self.req_dict.get(json_msg["transactionID"])
+                # get/delete request from dictionary
+                if (req is not None):
+                    with self.req_ops_lock:
+                        del self.req_dict[json_msg["transactionID"]]
+                    # put response into requester's reception queue
+                    log.debug("Msg:{0} dispatched to user queue".format(msg))
+                    req.user_queue.put(msg)
+                else:
+                    # TBD: it may be other messages, e.g., Actions, Armada Campaign
+                    log.info("Received unexpected message {0}".format(msg))
```

Comment:
Sorry for realizing this late. I think we can change this to ERROR or at least WARN, as we don't expect this to happen ever, and if it happens, probably something needs to be fixed.

Reply:
Fixed to WARN